### ISSUE TITLE
[HOLD for payment 2024-03-20] [$500] HIGH: Remove @expensify.sms from SMS mentions

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#33902]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.21-1
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** Applause - Internal Team
**Slack conversation:**

## Action Performed:
1. Navigate to staging.new.expensify.com
2. Go to a #Room that has several members (
3. Enter "@" in the compose box
4. Select phone number member

## Expected Result:
Phone number should be displayed in chat composer correctly

## Actual Result:
Phone number member is displayed with suffix @expensify.sms

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [x] Android: mWeb Chrome
- [x] iOS: Native
- [x] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

https://github.com/Expensify/App/assets/93399543/b2dd7d05-25cf-4ef8-b4a9-cab5ea78aad6

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0178aad4878bea5bbe</li>
        <li>Upwork Job ID: 1742631934961524736</li>
        <li>Last Price Increase: 2024-01-17</li>
<li>Automatic offers: </li>
<ul>
<li>c3024 | Reviewer | 28118032</li>
<li>FitseTLT | Contributor | 28118033</li>
</ul></li>
    </ul>
</details>

**Comment by FitseTLT** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Number contacts is displayed with suffix @expensify.sms

### What is the root cause of that problem?
We are pushing the login without removing the suffix here
https://github.com/Expensify/App/blob/27a51a44e744f66956687550768e0037e6612923/src/pages/home/report/ReportActionCompose/SuggestionMention.js#L166-L167
### What changes do you think we should make in order to solve the problem?

```
                    login: detail.login?.replace('@expensify.sms','')

```
### What alternative solutions did you explore? (Optional)

**Comment by namhihi237** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Phone number should be displayed in chat composer correctly
### What is the root cause of that problem?
We get `login` when select the phone number in the mention list here
https://github.com/Expensify/App/blob/3826e5dd01cde3ae14d2f258a05a9514fa11e158/src/pages/home/report/ReportActionCompose/SuggestionMention.js#L78
### What changes do you think we should make in order to solve the problem?
We should use LocalePhoneNumber.formatPhoneNumber() to remove sms domain with phone number

### What alternative solutions did you explore? (Optional)
N/A

**Comment by bernhardoj** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Selection of a user mention with a phone number will show "@expensify.com"

### What is the root cause of that problem?
When we select a user mention, we will take their raw login.
https://github.com/Expensify/App/blob/d4c3e63956e9c14838592172c7e0e8596b0814cf/src/pages/home/report/ReportActionCompose/SuggestionMention.js#L78

The mention object is constructed here:
https://github.com/Expensify/App/blob/d4c3e63956e9c14838592172c7e0e8596b0814cf/src/pages/home/report/ReportActionCompose/SuggestionMention.js#L163-L166

and the `login` property is coming from the personal detail object. If it's a phone number, the "@expensify.com" suffix will be present.

Previously in this [PR](https://github.com/Expensify/App/pull/31809), we already removed the "@expensify.com" suffix for `alternateText` property and we previously used `alternateText` when selecting a mention.
<img width="631" alt="image" src="https://github.com/Expensify/App/assets/50919443/7a6b11e1-6afc-4909-8bf1-8f9d410abb8b">

### What changes do you think we should make in order to solve the problem?

Take the mention object `alternateText` back.
```
`@${mentionObject.alternateText}`
```

**Comment by aim-salam** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Mention - Number contacts is displayed with suffix @expensify.sms

### What is the root cause of that problem?


`<TNodeChildrenRenderer tnode={props.tnode} />` render `props.tnode.init.textNode.data` that show full `+XXXXXX'@expensify'`
 

### What changes do you think we should make in order to solve the problem?

Change`props.tnode.init.textNode.data` by splitting at`@expensify` and acces first element

```js
props.tnode.init.textNode.data = displayNameOrLogin.split('@expensify')[0]
```

### What alternative solutions did you explore? (Optional)

**Comment by CortneyOfstad** [comment]
@c3024 any thoughts on the proposals above?

**Comment by c3024** [comment]
@FitseTLT @namhihi237 can you post the screenshot/video of how the sent mention messages appear? Are they pressable and does pressing on them show the details page of the mentioned user?

**Comment by namhihi237** [comment]
I think we should hold this issue utils here https://github.com/Expensify/App/issues/18729 to solve

**Comment by aim-salam** [comment]
Hello @c3024 . Mine have the screenshot video you requested. Here.
https://github.com/Expensify/App/issues/33902#issuecomment-1876676332

**Comment by CortneyOfstad** [comment]
> I think we should hold this issue utils here https://github.com/Expensify/App/issues/18729 to solve

@c3024 does that sound good to you? If not, can you please review the proposals above? Thanks!

**Comment by c3024** [comment]
I'll check today

**Comment by CortneyOfstad** [comment]
@c3024 any updates? Thanks!

**Comment by c3024** [comment]
Just got a clarification from [here](https://github.com/Expensify/App/issues/18729#issuecomment-1889967834).

@aim-salam can you see if you can remove the @expensify.sms in the composer as well and update your proposal accordingly?

**Comment by aim-salam** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Mention - Number contacts is displayed with suffix @expensify.sms

### What is the root cause of that problem?

In `SuggestionMention.js`, 
after phone number mentioned, `updateComment( )` show `mentionCode` as `@+XXXXXXX@expensify.sms`.
https://github.com/Expensify/App/blob/744e35c85157ecef9cdfab81a0ab56891dc507f9/src/pages/home/report/ReportActionCompose/SuggestionMention.js#L81

### What changes do you think we should make in order to solve the problem?

Remove `mentionCode`'s `@expensify.sms` into `''` whenever phone number mentioned.
```diff
//const into let
+            let mentionCode = mentionObject.text === CONST.AUTO_COMPLETE_SUGGESTER.HERE_TEXT ? CONST.AUTO_COMPLETE_SUGGESTER.HERE_TEXT : `@${mentionObject.login}`;
            const commentAfterMention = value.slice(suggestionValues.atSignIndex + suggestionValues.mentionPrefix.length + 1);

+            mentionCode = mentionCode.replace('@expensify.sms','')
```
But when we send the chat, the suggested number didn't mention.

![Screenshot 2024-01-13 at 1 22 03 PM](https://github.com/Expensify/App/assets/142758714/37b46079-6b09-45f2-a423-fd920b745d41)

In `reportactioncompose.js`, we need to add back '@expensify.sms' into phone number before we submit.

https://github.com/Expensify/App/blob/744e35c85157ecef9cdfab81a0ab56891dc507f9/src/pages/home/report/ReportActionCompose/ReportActionCompose.js#L279-L293

```diff
            const newComment = composerRef.current.prepareCommentAndResetComposer();
            if (!newComment) {
                return;
            }

+           const words = newComment.split(' ')
+           const updatedWords = words.map((word) => {
+                    if (word.charAt(0) === '@' && word.charAt(1) === '+') {
+                        return word + '@expensify.sms';
+                   } else {
+                       return word;
+                    }
+                })
+             const updatedComment = updatedWords.join(' ');

        
+            onSubmit(updatedComment);


```

https://github.com/Expensify/App/assets/142758714/1696c718-6b95-4b9e-b93d-dc450b23fe88

### What alternative solutions did you explore? (Optional)

**Comment by aim-salam** [comment]
@c3024 I misunderstood here; 
![Screenshot 2024-01-13 at 12 51 46 PM](https://github.com/Expensify/App/assets/142758714/cd011185-5f0f-4fc1-9c4d-d7db601cde4c)


I got confused by the video shown.
I proposed a new proposal https://github.com/Expensify/App/issues/33902#issuecomment-1890335529

**Comment by aim-salam** [comment]
@c3024 I misunderstood here;
I got confused by the video shown.
I proposed a new proposal https://github.com/Expensify/App/issues/33902#issuecomment-1890335529

**Comment by FitseTLT** [comment]
Updated [proposal](https://github.com/Expensify/App/issues/33902#issuecomment-1875886313)

**Comment by CortneyOfstad** [comment]
@c3024 any thoughts on the updated proposals above? Thanks!

**Comment by c3024** [comment]
I will update on the updated proposals today.

**Comment by c3024** [comment]
@puneetlath 

We are removing the @expensify.sms portion of the mentions in this issue.

We generally use `formatPhoneNumber` which uses `parsePhoneNumber` to remove the SMS domain. But for user mentions this is not feasible because this `parsePhoneNumber` breaks the phone number into parts separated by spaces like this (and country codes of same country numbers are also differently parsed based on IP with [this code here](https://github.com/Expensify/App/blob/991a4e213d6cb0af42b62b3e9f9c6b95b0317bea/src/libs/LocalePhoneNumber.ts#L34-L36))
<img width="366" alt="Screenshot 2024-01-19 at 1 30 43 PM" src="https://github.com/Expensify/App/assets/102477862/9c407641-6075-4434-b995-e84da56bda3b">
and mention strings cannot have spaces in between for proper parsing with `ExpensiMark` of these into mentions.

So we need to use the `Str.removeSMSDomain` to remove the sms domain and the result will be like this for mentions
<img width="148" alt="Screenshot 2024-01-19 at 1 36 06 PM" src="https://github.com/Expensify/App/assets/102477862/029e110e-b17e-4bf8-97be-5e3e7212219d">
and they look like these in composer
<img width="148" alt="Screenshot 2024-01-19 at 1 41 52 PM" src="https://github.com/Expensify/App/assets/102477862/e4e1e55f-97f5-4269-9616-42d718f707ea">


Can you please confirm if we can go ahead with parsing like these two screenshots above with `Str.removeSMSDomain`?

**Comment by CortneyOfstad** [comment]
@puneetlath bump on the comment above ^^^ thanks!

**Comment by puneetlath** [comment]
That sounds good to me.

**Comment by c3024** [comment]
[Proposal here](https://github.com/Expensify/App/issues/33902#issuecomment-1875886313) by @FitseTLT looks good to me.

🎀 👀 🎀 C+ Reviewed

**Comment by FitseTLT** [comment]
@jasperhuangg @c3024  This issue needs a change in our common library. So should I raise a PR and tag u guys on common lib repo??

**Comment by c3024** [comment]
Yes, a PR on `expensify-common` needs to be be raised and merged.

Then we bump the version in another PR along with other changes required on `Expensify/App`.

You can check a recent PR on `expensify-common`.

**Comment by FitseTLT** [comment]
Will create the PR on `expensify-common` in 2 days 👍

**Comment by CortneyOfstad** [comment]
Sounds good — thanks @FitseTLT!

**Comment by CortneyOfstad** [comment]
@FitseTLT any update on the PR? Thanks!

**Comment by CortneyOfstad** [comment]
@FitseTLT bump ^^^ thanks!

**Comment by CortneyOfstad** [comment]
Thanks @FitseTLT! @c3024 the PR is ready for review 👍 

@jasperhuangg — bump on my comment [here](https://github.com/Expensify/App/issues/33902#issuecomment-1932396959). Thanks!

**Comment by jasperhuangg** [comment]
@CortneyOfstad This looks like a #vip-vsb type of issue,j since it relates to mentions (which are not related to core expense features).

**Comment by CortneyOfstad** [comment]
Awesome — thanks @jasperhuangg!

**Comment by CortneyOfstad** [comment]
@c3024 any update on the PR review? Thanks!

**Comment by c3024** [comment]
Commented on the PR on expensify-common. Could you check it? @FitseTLT

**Comment by jasperhuangg** [comment]
Looking back at this issue, I think I might have misread what the issue was.

We need the @expensify.sms domain in the compose box when we send a mention since that's the only way we can identify a user, that is their login.

I thought we were talking about the domain existing in the report action after it is sent, but that's an entirely different issue. I believe we can just close this issue out without taking any action.

**Comment by c3024** [comment]
@jasperhuangg I had the same question if we should remove it in composer as well so I asked [here](https://github.com/Expensify/App/issues/18729#issuecomment-1888754799) and @puneetlath suggested [here](https://github.com/Expensify/App/issues/18729#issuecomment-1889967834) that we better remove it in composer as well.

If we would like to remove it only in the report action also, we can address it in this same issue.

**Comment by jasperhuangg** [comment]
Ah got it, that's my bad, I was under the impression that the expensify.sms domain was necessary in the composer. In that case let's keep this open, thanks for clarifying.

I spun up a revert of the linked PR since after testing it it didn't work.

**Comment by FitseTLT** [comment]
Created new PR with the test steps and also test result and tagged the reviewers 👍

**Comment by CortneyOfstad** [comment]
Awesome — thanks @FitseTLT!

**Comment by quinthar** [comment]
Woohoo, awesome progress.  What is the ETA for merge, and what specifically remains?  Who is doing it, when?

**Comment by FitseTLT** [comment]
On https://github.com/Expensify/expensify-common/pull/647 we need some backend job to be done in parallel to its change so @jasperhuangg is on it. As soon as it is finished, the change we need to make here won't take time 👍

**Comment by jasperhuangg** [comment]
I'm looking into the back-end fix today, I'll ping here when it's done. Looking to get the fix out for review today.

**Comment by CortneyOfstad** [comment]
Awesome — thanks @jasperhuangg!

**Comment by CortneyOfstad** [comment]
@jasperhuangg is there an update on this? Do you think this would be completed by today, EOW, etc? Thanks!

**Comment by jasperhuangg** [comment]
@CortneyOfstad ah sorry, I had some local changes but some other priorities came up, I'll be sure to get the changes polished up and ready for review today.

**Comment by jasperhuangg** [comment]
@FitseTLT @CortneyOfstad responded with some feedback in the PR! https://github.com/Expensify/expensify-common/pull/647#issuecomment-1955248478

**Comment by CortneyOfstad** [comment]
Looks to be good activity on the PR? Do you think that this is something that can be wrapped by EOD or by early next week @FitseTLT @jasperhuangg?

**Comment by CortneyOfstad** [comment]
Bump @FitseTLT @jasperhuangg^^^ thanks!

**Comment by FitseTLT** [comment]
Only left with @jasperhuangg final approval on the expensify-common PR 👍

**Comment by jasperhuangg** [comment]
@CortneyOfstad we merged the library fixes but need another App PR to wrap things up, which we should be able to get out early next week.

**Comment by CortneyOfstad** [comment]
Awesome — thank you!

**Comment by CortneyOfstad** [comment]
Not overdue per this comment [here](https://github.com/Expensify/App/issues/33902#issuecomment-1961894538) 👍

**Comment by jasperhuangg** [comment]
@FitseTLT any updates on that App PR to prevent rendering the expensify.sms domain with report actions?

**Comment by FitseTLT** [comment]
I think I can make it ready today or tomorrow

**Comment by jasperhuangg** [comment]
@quinthar We just merged the expensify-common PR that updates Expensimark.

@FitseTLT just needs to open another PR that ensures we don't display SMS domains in the front-end.

**Comment by CortneyOfstad** [comment]
Thanks @jasperhuangg!

@FitseTLT can you provide an update on when you can have the other PR opened by EOD?

**Comment by FitseTLT** [comment]
Yep

**Comment by quinthar** [comment]
@FitseTLT what's your ETA for a PR?

**Comment by FitseTLT** [comment]
PR is up and on reviewing stage

**Comment by CortneyOfstad** [comment]
Great! @c3024 ready for you!

**Comment by CortneyOfstad** [comment]
Hi @garrettmknight! I am heading OoO until March 11th, so reassigned this to keep it moving until I am back. As of right now, [the PR ](https://github.com/Expensify/App/pull/37559) from @FitseTLT is waiting for @c3024 to review. Thanks!

**Comment by c3024** [comment]
I've reviewed the PR and suggested some changes.

**Comment by FitseTLT** [comment]
I am asking for a response and confirmation from @jasperhuangg on the PR

**Comment by quinthar** [comment]
hi!  What exactly remains to be done, who's doing it, and what is the ETA?

**Comment by jasperhuangg** [comment]
I just responded to @FitseTLT's question on their PR. It's looking mostly good and we should be able to review and merge it soon.

**Comment by jasperhuangg** [comment]
We're waiting on @FitseTLT to fix some leftover review comments from @c3024

**Comment by quinthar** [comment]
Did this happen?  What is the holdup; what is the new ETA?

**Comment by FitseTLT** [comment]
@CortneyOfstad C+ review is done just waiting to be merged 👍

**Comment by quinthar** [comment]
What is the next step, who is doing it, and when?  Are we waiting on @c3024 to merge?  When can that happen?

**Comment by c3024** [comment]
There were conflicts and they have been fixed by @FitseTLT. The changes look good. @jasperhuangg will likely merge the PR by today.

**Comment by jasperhuangg** [comment]
Merged!

**Comment by c3024** [comment]
**BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:
 
* [x]  [@c3024] The PR that introduced the bug has been identified. Link to the PR: NA as this is a new feature.
* [x]  [@c3024] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment: NA
* [x]  [@c3024] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion: Not started because this is a new feature
* [x]  [@c3024] Determine if we should create a regression test for this bug. Yes
* [x]  [@c3024] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.

### Regression test proposal

This test in Tests of PR #33902 covers multiple cases. It is a good regression test.

>1. Go to any room and type @
>2. Select a user with phone number login
>3. Verify that the phone number inserted in the compose doesn't have the sms domain suffix @expensify.sms
>4. Send message and verify that the phone number is correct mention doesn't have the suffix and hover hover it and check the tooltip doesn't have it too
>5. Click on it and verify the profile page displays the phone number without the suffix
>6. On the system message click on invite the user to the room and on the message that indicates the user is invited check the tooltip displays without the suffix and click on it and verify the details page displays it without the suffix

**Comment by CortneyOfstad** [comment]
## Payment Summary

- @FitseTLT — paid $500 via Upwork
- @c3024 — pai $500 via Upwork

## Regression Test

https://github.com/Expensify/Expensify/issues/380809

Done and done! Thanks all!
